### PR TITLE
Document I/O interfaces

### PR DESCRIPTION
**Pull request by tmadlener**
BEGINRELEASENOTES
- Add docstrings to the `Reader` and `Writer` interface classes
- Make sure that `makeReader` throws an exception if multiple SIO files are passed (see [#647](https://github.com/AIDASoft/podio/issues/647))

ENDRELEASENOTES

### CODE REVIEW

File: include/podio/Reader.h [hunk 1]
```diff
@@ -73,50 +83,139 @@ class Reader {
...
+/// @note For SIO files this will only read the first file!
+///
+/// @param filenames The (paths to the) files to read from
+///
+/// @returns A Reader that has been initialized and that can be used for reading
+///          data from the passed files
+///
+/// @throws std::runtime_error in case the file extensions differ or in case
+///         support for the necessary I/O backend has not been built
 Reader makeReader(const std::vector<std::string>& filename);
```
Comment (hegner) [inline include/podio/Reader.h:219]: actually it should be `filenames`, not `filename`

File: include/podio/Reader.h [hunk 2]
```diff
@@ -73,50 +83,139 @@ class Reader {
...
 Reader makeReader(const std::string& filename);
+
+/// Create a Reader that is able to read the files
+///
+/// This will inspect the filenames as well as peek into the **first file only**
+/// to decide based on the contents which low level reader to instantiate for
+/// reading. All files are assumed to be of the same I/O format, no switching
+/// between formats is possible.
+///
+/// @note For SIO files this will only read the first file!
```
Comment (hegner) [inline include/podio/Reader.h:210]: we should put in some exception handling or something of that kind. That is too burried in the docs.

File: include/podio/Reader.h [hunk 3]
```diff
@@ -6,6 +6,16 @@
...
 namespace podio {
 
+/// Generic (type erased) reader class that can handle different I/O backends
+/// transparently
+///
+/// Offers some more high level functionality compared to the lower level
+/// backend specific readers that this class wraps. In contrast to the lower
+/// level readers that usually return arbitrary FrameData, this interface class
+/// will return fully constructed Frames.
+///
```
Comment (hegner) [inline include/podio/Reader.h:16]: In addition it provides convenience methods to deal specifically with the event frame category...

**Review by hegner — CHANGES_REQUESTED** [review]
Thanks a lot!
Just a few minor remarks. Except for the exception handling they could be covered in the same PR

File: include/podio/Reader.h (hunk 2 above)
Comment (tmadlener) [reply include/podio/Reader.h:210]: You mean, throw an exception when more than one SIO file is passed?

Comment (hegner) [reply include/podio/Reader.h:210]: we should not silently ignore faulty states.

Comment (hegner) [reply include/podio/Reader.h:210]: so yes

**Comment by tmadlener** [comment]
Thanks for checking. I think I addressed all of them.

**Review by hegner — APPROVED** [review]
That was quick! thanks

**Pull request merged**